### DISCUSSION TITLE
Support `firebase.js` along with `firebase.json`

### DISCUSSION THREAD

**Discussion by jeremytenjo** [firebase/firebase-js-sdk#5232] [General]
Supporting `firebase.js` would allow us to use global variables to set port numbers, output build directories, etc. 

For example `firebase.js`

```js
const { outputDir } = require('./constants')

...
hosting: {
    public: outputDir,
}
...
```

could share  `outputDir` with `webpack.config.js`

```js
const { outputDir } = require('./constants')

...
{
path: outputDir
}
...
```

**Comment by shivamCode0** [comment]
Isn't this a part of [firebase-tools](https://github.com/firebase/firebase-tools)?

**Comment by Feiyang1** [comment]
@jeremytenjo Can you provide more detail to help me understand your use case? Setting output directory sounds like the job of the build tools, not the JS SDK. Do you mean something else?

**Reply by jeremytenjo** [reply]
You are right, i'll take this idea to [firebase-tools](https://github.com/firebase/firebase-tools)

**Comment by akauppi** [comment]
As Feiyang wrote, this is in the wrong section - but it's easy to be confused with the JS SDK vs. CLI differences, at first.

A brief reply to your question, though. What you're suggesting is good, but it's not hard even now to read in `firebase.json` to one's build scripts. I do it all the time - either with `import` (ESM) or with two lines of Node code: 

```
const raw = readFileSync(firebaseJson);
const conf = JSON.parse(raw);
```

You need to be careful whether expected fields really are there - but that would be the same with `firebase.js` just as well.

As a fellow user of Firebase, I would not advocate a change, since it adds complexity to the Firebase (CLI) code base.

**Reply by shivamCode0** [reply]
Yes, I also agree that it should not change. This sounds similar to [this next.js issue](https://github.com/vercel/next.js/issues/5318) where someone wanted to use a TypeScript config. In both cases, the best option is to not introduce complexity.
